File: ProjectTest/OneLang/CSharp/One/Ast/AstHelper.cs
using System.Collections.Generic;
using One.Ast;

namespace One.Ast
{
    public class AstHelper {
        public static IInterface[] collectAllBaseInterfaces(IInterface intf) {
            var result = new Set<IInterface>();
            var toBeProcessed = new List<IInterface> { intf };
            
            while (toBeProcessed.length() > 0) {
                var curr = toBeProcessed.pop();
                result.add(curr);
                
                if (curr is Class && ((Class)curr).baseClass != null)
                    toBeProcessed.push((((ClassType)((Class)curr).baseClass)).decl);
                
                foreach (var baseIntf in curr.baseInterfaces)
                    toBeProcessed.push((((InterfaceType)baseIntf)).decl);
            }
            
            return Array.from(result.values());
        }
    }
}

File: ProjectTest/OneLang/Java/src/main/java/OneLang/One/Ast/AstHelper/AstHelper.java
import java.util.HashSet;
import java.util.List;
import java.util.ArrayList;

public class AstHelper {
    public static IInterface[] collectAllBaseInterfaces(IInterface intf) {
        var result = new HashSet<IInterface>();
        var toBeProcessed = new ArrayList<>(List.of(intf));
        
        while (toBeProcessed.size() > 0) {
            var curr = toBeProcessed.remove(toBeProcessed.size() - 1);
            result.add(curr);
            
            if (curr instanceof Class && ((Class)curr).baseClass != null)
                toBeProcessed.add((((ClassType)((Class)curr).baseClass)).decl);
            
            for (var baseIntf : curr.getBaseInterfaces())
                toBeProcessed.add((((InterfaceType)baseIntf)).decl);
        }
        
        return result.toArray(IInterface[]::new);
    }
}